### PR TITLE
Add Ballerina Interceptor Implementation

### PR DESCRIPTION
**Pull request by hasithaa**
Guide - https://docs.google.com/document/d/1ShCvpDb6NhWg13FFkQDQOhg8X3Cwn5F3Kv5QnnYzq-w

Please merge https://github.com/ballerinalang/ballerina-parent/pull/15 before merging this.

Dependent PRs.
https://github.com/ballerinalang/tools-distribution/pull/107

### CODE REVIEW

File: modules/ballerina-core/src/main/java/org/ballerinalang/runtime/ServerConnectorMessageHandler.java [hunk 1]
```diff
@@ -90,11 +95,40 @@ public static void handleInbound(CarbonMessage cMsg, CarbonCallback callback) {
 
             // Find the Resource
             ResourceInfo resource = resourceDispatcher.findResource(service, cMsg, callback);
-
-            invokeResource(cMsg, callback, resource, service);
-
+            // engage Service interceptors.
+            CarbonCallback resourceCallback = callback;
+            if (serviceInterceptors.isEnabled(protocol)) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/runtime/ServerConnectorMessageHandler.java]: It is better to move this code block to a method in this class.

File: modules/ballerina-core/src/main/java/org/ballerinalang/runtime/ServerConnectorMessageHandler.java [hunk 2]
```diff
@@ -90,11 +95,40 @@ public static void handleInbound(CarbonMessage cMsg, CarbonCallback callback) {
...
+                ServerConnectorInterceptorInfo interceptorInfo = serviceInterceptors
+                        .getServerConnectorInterceptorInfo(protocol);
+
+                if (interceptorInfo != null) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/runtime/ServerConnectorMessageHandler.java]: Can the 'interceptorInfo' be null? If not, we can remove this check

File: modules/ballerina-core/src/main/java/org/ballerinalang/runtime/config/BLangConfigurationManager.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import org.wso2.carbon.config.ConfigProviderFactory;
+import org.wso2.carbon.config.ConfigurationException;
+import org.wso2.carbon.config.provider.ConfigProvider;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+/**
+ * Utility methods for loading Ballerina Configuration.
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/runtime/config/BLangConfigurationManager.java]: Can you add the @Since doc annotation

**Comment by sameera-jayasoma** [comment]
Can you please explain the high-level implementation design with the following classes in your PR?

BLangServiceInterceptors.java
BLangVMInterceptors.java
ResourceInterceptor.java
ResourceInterceptorCallback.java
ServerConnectorInterceptorInfo.java

File: modules/ballerina-core/src/main/java/org/ballerinalang/runtime/ServerConnectorMessageHandler.java (hunk 1 above)
Comment (hasithaa) [inline modules/ballerina-core/src/main/java/org/ballerinalang/runtime/ServerConnectorMessageHandler.java]: Yes, we can make this to static method, which returns Carboncallback, but this requires additional null check. This callback can be null either result.getMessageIntercepted() == null or !result.isInvokeNext().

If callback is null then we should return from the function.

File: modules/ballerina-core/src/main/java/org/ballerinalang/runtime/config/BLangConfigurationManager.java (hunk 3 above)
Comment (hasithaa) [inline modules/ballerina-core/src/main/java/org/ballerinalang/runtime/config/BLangConfigurationManager.java]: Added.

File: modules/ballerina-core/src/main/java/org/ballerinalang/runtime/ServerConnectorMessageHandler.java (hunk 2 above)
Comment (hasithaa) [inline modules/ballerina-core/src/main/java/org/ballerinalang/runtime/ServerConnectorMessageHandler.java]: Removed.

**Comment by hasithaa** [comment]
- ResourceInterceptor.java
  This is a data holder class for Resource Interceptor, contains information about its programFile, packageInfo, and FunctionInfo. 

- ServerConnectorInterceptorInfo.java
This a data holder class containing all details about the message interception for a particular ServerConnector.  i.e request/response chains (List of ResourceInterceptor objects) . deployment directory etc. 

- BLangServiceInterceptors.java
This class responsible for initializing interceptors in Ballerina runtime. This class reads deployment.yaml and build ServerConnectorInterceptorInfo objects for each server Connector type defined in deployment.yaml 

-BLangVMInterceptors.java
 This is utility class for invoking an interceptor using Ballerina VM. This class contains an util function, to invoke a ResourceInterceptor with given message. It work same as BLangFunctions.invokeNew() but this is optimized for interception invocation by pre allocating caller's stack etc. 

- ResourceInterceptorCallback.java
This is wrapper Carbon callback which wraps client's callback. It invokes response interceptors before it calls client's callback.

File: modules/ballerina-core/src/main/java/org/ballerinalang/runtime/interceptors/BLangVMInterceptors.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import org.ballerinalang.util.codegen.WorkerInfo;
+import org.ballerinalang.util.exceptions.BLangRuntimeException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+
+/**
+ * BLangVMInterceptors is utility class for invoking an interceptor using Ballerina VM.
+ */
```
Comment (lankavitharana) [inline modules/ballerina-core/src/main/java/org/ballerinalang/runtime/interceptors/BLangVMInterceptors.java]: Can you add @Since annotation here as well?

**Pull request merged**